### ISSUE TITLE
<ServerGarbageCollection>false</ServerGarbageCollection> does not set GC to Client mode on Windows Server 2016

### ISSUE THREAD

**Issue by NickCraver** [dotnet/core-setup#5551]
While gathering info for https://github.com/aspnet/AspNetCore/issues/3409, I was getting allocation thresholds for GC collection under various configurations.

On Windows 10, the default GC mode is "Client". Forcing server via `<ServerGarbageCollection>true</ServerGarbageCollection>` in the `.csproj` works (in both `net472` and `netcoreapp2.1`). However, on Windows Server 2106 the default GC mode is "Server". While a way exists to turn this *on*, there doesn't seem to be an equivalent mechanism to turn it *off*.

Example `.csproj`:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472;netcoreapp2.1</TargetFramework>
    <ServerGarbageCollection>false</ServerGarbageCollection>
  </PropertyGroup>
</Project>
```

The lack of any effect on GC was verified at runtime via:
```c#
using System;
using System.Runtime;
using static System.Console;

namespace AllocationsTest
{
    public static class Program
    {
        public static void Main(string[] args)
        {
			WriteLine("  GC Mode: " + (GCSettings.IsServerGC ? "Server" : "Client"));
			ReadKey();
		}
	}
}
```

What does work is an environmental variable approach:
```
SET COMPlus_gcServer=0
```
...but it's my understanding this should work from the `.csproj`/MSBuild properties as well.

The flip side of this is `<ServerGarbageCollection>true</ServerGarbageCollection>` to force GC into server mode. The result of this is a section of `*.runtimeconfig.json`:
```json
{
  "runtimeOptions": {
    "tfm": "netcoreapp2.1",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "2.1.0"
    },
    "configProperties": { 
      "System.GC.Server": true 
    }
  }
}
```
When the `ServerGarbageCollection` MSBuild property is false, the section isn't generated at all. Letting it generate with `true` (in a publish) and manually flipping it to `false` for running that publish also had no effect - still server mode.

In .NET 4.7.2, the old `app.config` approach also works (but has no effect on `netcoreapp2.1`, as expected):
```xml
<configuration><runtime><gcServer enabled="false" /></runtime></configuration>
```

Note for anyone looking: **this does not happen on a Client OS**. It requires a server OS's default behavior to reproduce.

cc @sebastienros @Maoni0 @adamsitnik @davidfowl

**Comment by jaredpar** [comment]
@NickCraver done.

**Comment by vitek-karas** [comment]
This is a combination of various bugs and unexpected behaviors.

First, I tried this on Windows Server 2016 (vanilla Data Center install with VS) and the default GC mode was Client. Not sure if there's some registry setting which affects this. @Maoni0  what is the expected default on Windows Server 2016 (for .NET Core 2.1 and 3.0)?

### .NET Core 2.*
`dotnet build` and `dotnet run` produce a somewhat hacked up output which doesn't contain everything it should. One of the effects is that runtime properties are not populated in `.runtimeconfig.json`. I filed dotnet/sdk#3043 for this. Not sure if we can retroactively fix it for 2.*.

`dotnet publish` seems to work fine, it puts the property into the `.runtimeconfig.json` as it should.

Running the published app with
```
dotnet run .\bin\Release\netcoreapp2.1\publish -c Release
```
doesn't actually do what one would expect. The command line parser recognizes the `-c Release` and uses that to run a release build, but it doesn't look for any "non option" and treats anything it doesn't understand as arguments for the app. So it passes the `.\bin\Release\netcoreapp2.1\publish` as a first argument to the application.

To run the published app, use
```
dotnet .\bin\Release\netcoreapp2.1\public\app.dll
```

### .NET Core 3.0
In 3.0 the `dotnet build` is very similar to `dotnet publish` and so it correctly produces the `.runtimeconfig.json`. This also means that simple `dotnet run` works as well.

Unfortunately it's not all perfect either as there's still a bug in incremental build where `.runtimeconfig.json` is not updated even when it should be. This might be another one participating in the unexpected behavior: dotnet/sdk#2787.

**Comment by Maoni0** [comment]
@vitek-karas thanks for such detailed analysis! 

>First, I tried this on Windows Server 2016 (vanilla Data Center install with VS) and the default GC mode was Client. Not sure if there's some registry setting which affects this. @Maoni0 what is the expected default on Windows Server 2016 (for .NET Core 2.1 and 3.0)?

just to clarify - by "tried" you mean just running an app without specifying <ServerGarbageCollection> and this app is NOT an asp.net core app or anything else that's supposed to turn on Server GC for you, correct? if that's the case it should always be Workstation GC.

**Comment by vitek-karas** [comment]
@Maoni0 yes - I tried a vanilla `dotnet new console` so that explains it.

**Comment by nguerrera** [comment]
> .NET Core 2.*
>
> One of the effects is that runtime properties are not populated in .runtimeconfig.json. I filed dotnet/sdk#3043 for this. Not sure if we can retroactively fix it for 2.*.

I have tried 5 different SDK versions and cannot reproduce this: https://github.com/dotnet/sdk/issues/3043#issuecomment-477121324

**Comment by vitek-karas** [comment]
> .NET Core 2.*
> 
> One of the effects is that runtime properties are not populated in .runtimeconfig.json.

I'm sorry - I can't repro this either now. Not sure how I came to the conclusion yesterday. Right now all I see is the problem with the incremental build.

**Comment by nguerrera** [comment]
> Right now all I see is the problem with the incremental build.

If that's what we believe is the only problem, we can close this as a duplicate of dotnet/sdk#2787.
